Finally, I debugged it today. The problem is in PNP BIOS. pnp_bus_suspend() 
calls pnp_stop_dev() for the device if the device can be disabled according 
to pnp_can_disable(). The problem is that pnpbios_disable_resources() 
returns -EPERM if the device is not dynamic (!pnpbios_is_dynamic()) but 
insert_device() happily sets PNP_DISABLE capability/flag even if the device 
is not dynamic. So we try to disable non-dynamic devices which will fail. 
This patch prevents insert_device() from setting PNP_DISABLE if the device is 
not dynamic and fixes suspend on my system.

Signed-off-by: Ondrej Zary <linux@rainbow-software.org>

--- linux-2.6.17.orig/drivers/pnp/pnpbios/core.c	2006-08-15 18:12:15.129352250 +0200
+++ linux-2.6.17/drivers/pnp/pnpbios/core.c	2006-08-15 18:12:36.366679500 +0200
@@ -346,7 +346,7 @@
 	dev->flags = node->flags;
 	if (!(dev->flags & PNPBIOS_NO_CONFIG))
 		dev->capabilities |= PNP_CONFIGURABLE;
-	if (!(dev->flags & PNPBIOS_NO_DISABLE))
+	if (!(dev->flags & PNPBIOS_NO_DISABLE) && pnpbios_is_dynamic(dev))
 		dev->capabilities |= PNP_DISABLE;
 	dev->capabilities |= PNP_READ;
 	if (pnpbios_is_dynamic(dev))
